language/move-binary-format/src/compatibility.rs

```diff
@@ -156,11 +156,19 @@ impl Compatibility {
             {
                 // if it was public(script), it must remain public(script)
                 // if it was not public(script), it _cannot_ become public(script)
-                old_func.is_entry == new_func.is_entry
+                // if it was private before, it can become public(script)
+                old_func.visibility == Visibility::Private || old_func.is_entry == new_func.is_entry
             } else {
-                // If it was an entry function, it must remain one.
                 // If it was not an entry function, it is allowed to become one.
-                !old_func.is_entry || new_func.is_entry
+                // If it was a private entry function and we have `check_private_entry_linking` then it must remain one
+                // If it was a public entry function, it must remain one.
```

Comment:
I would decouple `public` and  `entry` here -- i.e. we should not have a different rule for `public entry` to just `entry`, it should just be that if we have `check_private_entry_linking` enabled then a function cannot lose its `entry` modifier.

Reply:
So the rule should be something like below, although it might be better off in another place:
```
if old_func.is_entry && !new_func.is_entry {
    private_entry_linking = false
}
```

(Possible you want to rename it to just `entry_linking` given that it is truly orthogonal to `public` vs `(private)`).

Reply:
Much prefer this. Thank you for the suggestion!